relayer/paths/path_processor.go

```diff
@@ -0,0 +1,499 @@
+	pp.pathEnd2.chainProcessor = chainProcessor
+}
+
+func (pp *PathProcessor) HandleNewMessages(chainID string, channelKey ibc.ChannelKey, messages map[string]map[uint64]provider.RelayerMessage) {
+	if pp.pathEnd1.info.ChainID == chainID {
+		if pp.pathEnd1.info.ShouldRelayChannel(channelKey) {
+			pp.handleNewMessagesForPathEnd(pp.pathEnd1, messages)
+		}
+	} else if pp.pathEnd2.info.ChainID == chainID {
+		if pp.pathEnd2.info.ShouldRelayChannel(channelKey) {
+			pp.handleNewMessagesForPathEnd(pp.pathEnd2, messages)
+		}
+	}
+}
+
+func (pp *PathProcessor) handleNewMessagesForPathEnd(
+	p *PathEndRuntime,
+	newMessages map[string]map[uint64]provider.RelayerMessage,
+) {
+	// fmt.Printf("{%s} New messages: +%v\n", p.info.ChainID, newMessages)
+	p.messagesLock.Lock()
+	for msg, sequencesMap := range newMessages {
+		if len(sequencesMap) == 0 {
+			continue
+		}
+		for sequence, ibcMsg := range sequencesMap {
+			// update max sequence for MsgTransfer messages
+			if msg == ibc.MsgTransfer {
+				p.maxSequenceLock.Lock()
+				p.updateMaxSequence(sequence)
+				p.maxSequenceLock.Unlock()
+			}
+			if _, ok := p.messages[msg]; !ok {
+				p.messages[msg] = make(map[uint64]provider.RelayerMessage)
+			}
+			p.messages[msg][sequence] = ibcMsg
+		}
+	}
+	p.messagesLock.Unlock()
+	// only start processing packets if querying latest heights for both chains
+	if pp.pathEnd1.chainProcessor.InSync() && pp.pathEnd2.chainProcessor.InSync() {
+		pp.ScheduleNextProcess(true)
+	}
+}
+
+func (pp *PathProcessor) IsRelevantChannel(chainID string, channelKey ibc.ChannelKey) bool {
+	if pp.pathEnd1.info.ChainID == chainID {
+		return pp.pathEnd1.info.ShouldRelayChannel(channelKey)
+	} else if pp.pathEnd2.info.ChainID == chainID {
+		return pp.pathEnd2.info.ShouldRelayChannel(channelKey)
+	}
+	return false
+}
+
+func (pp *PathProcessor) IsRelevantClient(clientID string) bool {
+	return pp.pathEnd1.info.ClientID == clientID || pp.pathEnd2.info.ClientID == clientID
+}
+
+func (pp *PathProcessor) GetRelevantConnectionID(chainID string) (string, error) {
+	if pp.pathEnd1.info.ChainID == chainID {
+		return pp.pathEnd1.info.ConnectionID, nil
+	} else if pp.pathEnd2.info.ChainID == chainID {
+		return pp.pathEnd2.info.ConnectionID, nil
+	}
+	return "", errors.New("irrelevant chain processor provided")
+}
+
+// gets unrelayed packets and acks and cleans up sequence retention for completed packet flows
+func (pp *PathProcessor) getUnrelayedPacketsAndAcks(srcPathEnd, dstPathEnd *PathEndRuntime) ([]ibc.IBCMessageWithSequence, []ibc.IBCMessageWithSequence) {
+	srcPathEnd.messagesLock.Lock()
+	dstPathEnd.messagesLock.Lock()
+	srcPathEnd.inProgressMessageSendsLock.Lock()
+	dstPathEnd.inProgressMessageSendsLock.Lock()
+	defer srcPathEnd.messagesLock.Unlock()
+	defer dstPathEnd.messagesLock.Unlock()
+	defer srcPathEnd.inProgressMessageSendsLock.Unlock()
+	defer dstPathEnd.inProgressMessageSendsLock.Unlock()
```

Comment:
Taking several locks in sequence will incur a high risk of deadlock, for any application that has enough goroutines active.

Reply:
I missed this somehow. Yeah, high chance of deadlock. As Mark stated:

>  A single goroutine is responsible for all of the mutable data, without requiring any mutexes.

I am also a fan of this pattern. I've heard it called a "single writer" pattern. Only 1 goroutine writes (or mutates) data. 

To echo Mark, hard to tell if that's feasible here though.

Reply:
I cleaned this section up by making copies of the messages, using that for iteration to determine what needs to be deleted, then only locking/deleting for the specific maps after all of that.

I also removed the locks for the ibc message assembly, since those can just be separate slices that are concatenated after the waitgroup.

Right now the chain processors push messages into the path processor, and the path processor manages the backlog of messages coming from both chains, and clears out old packets. That data could be separated into two different stores, but we'd still have to lock for read/writes since the data is being read/written by different threads.